2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+If the local cache does not exist, the cmdlet will call `Update-PSResourceCache`
+first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules.
+A `-AllowPrerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name.
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-AllowPrerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-AllowUntrusted` can be used to suppress being prompted for untrusted sources.
+`-AllowDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-AllowReinstall` can be used to re-install a module.
+
+A `-Quiet` switch will suppress progress information.
+
+### Dependencies and version management
+
+`Install-PSResource` can accept a path to a psd1 file (using `-RequiredModulesFile`)
+or a hashtable (using `-RequiredModules`) where the key is the module name and the
+value is either the required version specified using Nuget version range syntax or
+a hash table where `repository` is set to the URL of the repository and
+`version` contains the [Nuget version range syntax](https://docs.microsoft.com/en-us/nuget/reference/package-versioning#version-ranges-and-wildcards).
+
+```powershell
+Install-PSResource -RequiredModules @{
+  "Configuration" = "[1.3.1,2.0)"
+  "Pester"        = @{
+    version = "[4.4.2,4.7.0]"
+    repository = "https://www.powershellgallery.com"
+  }
+}
+```
+
+### Saving resources
+
+With the removal of PackageManagement, there is still a need to support saving
+arbitrary nupkgs (assemblies) used for scripts.
+
+`Save-PSResource -Type Library` will download nupkgs that have a `lib` folder.
+A `-AllowPrerelease` switch allows saving prerelease versions.
+The dependent native library in `runtimes` matching the current system runtime
+will be copied to the same destination.
+
+This cmdlet has the same parameters as `Install-PSResource` with the addition
+of mandatory `-Path` or `-LiteralPath` parameters.
+
+### Updating resources
+
+`Update-PSResource` will update all resources to most recent minor version by
+default.
+A `-OnlyMinorUpdates` switch will allow only updating to newer minor version.
```

Comment:
Can we make this the reverse so it only updates minor by default and you have to opt in to update major versions? The PS world isn't great at following Semver (or similar style) where major version changes indicate breaking changes but hopefully we'll get better at that and we should encourage it as much as possible to follow that model.

Reply:
I debated on this one.  Not sure which way to go is the ideal case.  Originally I had a `-AllowMajorVersionUpdates`, but decided that I believe most modules don't follow semver, so a major version is really a minor version by a different name so went with `-OnlyMinorUpdates` so that by default you get major version updates.  Would like some more input on which way to go here.

Reply:
I can imagine being frustrated running `Update-Module` on some module and thinking it's up to date, only to find out later that without a flag it didn't actually give me the latest update

Reply:
If this were globally configurable it might offer a reasonable compromise. But I do hesitate to suggest a global configuration.

This version granularity question seems to recur in a few places.

My personal ideal would be something like `Update-Module -AllowedVersionChange Major` (need to workshop that parameter name).

That could then either default to `Major` or refer to a configuration (that itself defaults to `Major`). I think a global configuration could be difficult to manage properly, but would also be pretty convenient and suit many people's needs

Reply:
I think it should default to minor (compatible feature updates), and write warnings about skipped updates, telling you the switch you need to apply them.

I also think it should have a switch for `-BugFixOnly` to support restricting it to only new patch releases, for those who are more cautious.

Reply:
@kilasuit proposed the Major/Minor thing in PowerShellGet 2.x. I wasn't too enthusiastic, and TBH I am still not. You can see the discussion at https://github.com/PowerShell/PowerShellGet/issues/422

Reply:
I don't think we should let the fact that most modules don't follow SemVer have an impact on this. We should be pushing for best practices and showing how things should be done from MSFT commands/modules, if a version doesn't include breaking changes then it's a minor update and should be updated to without issues, if a version includes breaking changes then it's a major version change and should be opt in. 

Having a warning show up when you've installed a minor version but there are major versions available would be useful, it lets the user know that they could upgrade further but there might be problems with it (and the warning should include text about major version changes possibly including breaking changes).

As a module user I want to know when updating a module that I've got the latest version that works with my current scripts/modules. For me that's easy enough to do as I know I should set -MaximumVersion to 1.9999.99999 or something similar to ensure I don't get any major version changes (with the assumption that they are breaking changes), but the average PS user won't know to do that (or possibly even know that major version = possible breaking changes) so we need to account for that. I will admit it's also an issue with educating users and developers about version numbers and how to do them right but that's outside the scope of this RFC, the best we can do here is to lead the way.

Reply:
I'm in agreement with @ChrisLGardner here, over time the landscape of developers will change and us being more cautious here by only jumping up minor versions by default plays it the safest overall. I also like @Jaykul idea of `-BugFixOnly` as this would allow scripters to have builds that auto update dependencies (akin to dependabot) in a part of their pipeline and run their own tests to confirm / deny breakages between versions, though that's a long way away but this should be the direction to think about.

Reply:
Going with `-AllowMajorVersionUpdate` and `-PatchUpdateOnly`

Reply:
What happens when you pass both `-AllowMajorVersionUpdate` and `-PathUpdateOnly` or will they be in different parameter sets?

I like @rjmholt idea of using a single parameter `-MaximumVersionUpdate` with an enum `Major`, `Minor`, `Patch`.